tests/data_group_test.py

```diff
@@ -1,8 +1,200 @@
+import pytest
+
 import scipp as sc
 
 
+def test_create_from_dict_works_with_mixed_types_and_non_scipp_objects():
+    items = {
+        'a': 1,
+        'b': sc.scalar(2),
+        'c': sc.DataArray(sc.scalar(3)),
+        'd': np.arange(4)
+    }
+    dg = sc.DataGroup(items)
+    assert tuple(dg.keys()) == ('a', 'b', 'c', 'd')
+
+
+def test_len():
+    assert len(sc.DataGroup()) == 0
+    assert len(sc.DataGroup({'a': 0})) == 1
+    assert len(sc.DataGroup({'a': 0, 'b': 1})) == 2
+
+
+def test_iter_empty():
+    dg = sc.DataGroup()
+    it = iter(dg)
+    with pytest.raises(StopIteration):
+        next(it)
+
+
+def test_iter_nonempty():
+    dg = sc.DataGroup({'a': sc.scalar(1), 'b': sc.scalar(2)})
+    it = iter(dg)
+    assert next(it) == 'a'
+    assert next(it) == 'b'
+    with pytest.raises(StopIteration):
+        next(it)
+
+
+def test_getitem_str_key():
+    dg = sc.DataGroup({'a': sc.scalar(1), 'b': sc.scalar(2)})
+    assert sc.identical(dg['a'], sc.scalar(1))
+    assert sc.identical(dg['b'], sc.scalar(2))
+
+
+def test_setitem_str_key_adds_new():
+    dg = sc.DataGroup()
+    dg['a'] = sc.scalar(1)
+    assert 'a' in dg
+
+
+def test_setitem_str_key_replace():
+    dg = sc.DataGroup({'a': sc.scalar(1)})
+    dg['a'] = sc.scalar(2)
+    assert sc.identical(dg['a'], sc.scalar(2))
+
+
+def test_delitem_removes_item():
+    dg = sc.DataGroup({'a': sc.scalar(1)})
+    del dg['a']
+    assert 'a' not in dg
+
+
+def test_delitem_raises_KeyError_when_not_found():
+    dg = sc.DataGroup({'a': sc.scalar(1)})
+    with pytest.raises(KeyError):
+        del dg['b']
+
+
+def test_dims_with_scipp_objects_combines_dims_in_insertion_order():
+    assert sc.DataGroup({'a': sc.scalar(1)}).dims == ()
+    assert sc.DataGroup({'a': sc.ones(dims=('x', ), shape=(2, ))}).dims == ('x', )
+    assert sc.DataGroup({
+        'a': sc.ones(dims=('x', ), shape=(2, )),
+        'b': sc.ones(dims=('y', ), shape=(2, ))
+    }).dims == ('x', 'y')
+    assert sc.DataGroup({
+        'b': sc.ones(dims=('y', ), shape=(2, )),
+        'a': sc.ones(dims=('x', ), shape=(2, ))
+    }).dims == ('y', 'x')
+    # In this case there would be a better order, but in general there is not, so at
+    # leasr for now the implementation makes no attempt for this.
+    assert sc.DataGroup({
+        'a': sc.ones(dims=('x', 'z'), shape=(2, 2)),
+        'b': sc.ones(dims=('y', 'z'), shape=(2, 2))
+    }).dims == ('x', 'z', 'y')
+
+
+def test_ndim():
+    assert sc.DataGroup({'a': sc.scalar(1)}).ndim == 0
+    assert sc.DataGroup({'a': sc.ones(dims=('x', ), shape=(1, ))}).ndim == 1
+    assert sc.DataGroup({'a': sc.ones(dims=('x', 'y'), shape=(1, 1))}).ndim == 2
+    assert sc.DataGroup({
+        'a': sc.ones(dims=('x', ), shape=(2, )),
+        'b': sc.ones(dims=('y', ), shape=(2, ))
+    }).ndim == 2
+
+
+def test_non_scipp_objects_are_considered_to_have_0_dims():
+    assert sc.DataGroup({'a': np.arange(4)}).dims == ()
+    assert sc.DataGroup({'a': np.arange(4)}).ndim == 0
```

Comment:
What happens in operations with such a data group? Would it, e.g., add numpy arrays together in `dg1+dg2`?

Reply:
Yes, it should work, since the implementation uses `operation.add`. Test for binary ops (and more binary ops) will be addressed in a follow-up PR.

Reply:
I also have a question about items in data groups that cannot be added, but I will wait for the follow-up PR.